presto-base-arrow-flight/src/main/java/com/facebook/plugin/arrow/AbstractArrowMetadata.java

```diff
@@ -0,0 +1,269 @@
+
+    protected abstract String getDataSourceSpecificTableName(ArrowFlightConfig config, String tableName);
+
+    @Override
+    public ConnectorTableHandle getTableHandle(ConnectorSession session, SchemaTableName tableName)
+    {
+        if (!listSchemaNames(session).contains(tableName.getSchemaName())) {
+            return null;
+        }
+
+        if (!listTables(session, Optional.ofNullable(tableName.getSchemaName())).contains(tableName)) {
+            return null;
+        }
+        return new ArrowTableHandle(tableName.getSchemaName(), tableName.getTableName());
+    }
+
+    public List<Field> getColumnsList(String schema, String table, ConnectorSession connectorSession)
+    {
+        try {
+            String dataSourceSpecificSchemaName = getDataSourceSpecificSchemaName(config, schema);
+            String dataSourceSpecificTableName = getDataSourceSpecificTableName(config, table);
+            FlightDescriptor flightDescriptor = getFlightDescriptor(Optional.empty(),
+                    dataSourceSpecificSchemaName, dataSourceSpecificTableName);
+
+            Optional<Schema> flightschema = clientHandler.getSchema(flightDescriptor, connectorSession);
+            List<Field> fields = flightschema.map(Schema::getFields).orElse(Collections.emptyList());
+            return fields;
+        }
+        catch (Exception e) {
+            throw new ArrowException(ARROW_FLIGHT_ERROR, "The table columns could not be listed for the table " + table, e);
+        }
+    }
+
+    @Override
+    public Map<String, ColumnHandle> getColumnHandles(ConnectorSession session, ConnectorTableHandle tableHandle)
+    {
+        Map<String, ColumnHandle> columnHandles = new HashMap<>();
+
+        String schemaValue = ((ArrowTableHandle) tableHandle).getSchema();
+        String tableValue = ((ArrowTableHandle) tableHandle).getTable();
+        String dbSpecificSchemaValue = getDataSourceSpecificSchemaName(config, schemaValue);
+        String dBSpecificTableName = getDataSourceSpecificTableName(config, tableValue);
+        List<Field> columnList = getColumnsList(dbSpecificSchemaValue, dBSpecificTableName, session);
+
+        for (Field field : columnList) {
+            String columnName = field.getName();
+            logger.debug("The value of the flight columnName is:- %s", columnName);
+
+            Type type = getPrestoTypeFromArrowField(field);
+            columnHandles.put(columnName, new ArrowColumnHandle(columnName, type));
+        }
+        return columnHandles;
+    }
+
+    @Override
+    public List<ConnectorTableLayoutResult> getTableLayouts(ConnectorSession session, ConnectorTableHandle table, Constraint<ColumnHandle> constraint, Optional<Set<ColumnHandle>> desiredColumns)
+    {
+        if (!(table instanceof ArrowTableHandle)) {
+            throw new PrestoException(
+                    StandardErrorCode.INVALID_CAST_ARGUMENT,
```

Comment:
In this case just use `checkArgument`, it's a precondition failure.  Internal error is OK.

Reply:
Updated to use checkArgument